fpm/src/fpm_backend.f90

```diff
@@ -120,4 +119,41 @@ recursive subroutine build_source(model,source_file,linking)
 
 end subroutine build_source
 
+
+function get_object_name(model,source_file_name) result(object_file)
+    ! Generate object target path from source name and model params
+    !  
+    !  src/test.f90        ->  <output-dir>/<package-name>/test.o
+    !  src/subdir/test.f90 ->  <output-dir>/<package-name>/subdir_test.o
+    !
+    type(fpm_model_t), intent(in) :: model
+    character(*), intent(in) :: source_file_name
+    character(:), allocatable :: object_file
+
+    integer :: i
+    character(1) :: filesep
+
+    select case(get_os_type())
+    case (OS_WINDOWS)
+        filesep = '\'
+    case default
+        filesep = '/'
+    end select
+
+    ! Exclude first directory level from path
+    object_file = source_file_name(index(source_file_name,filesep)+1:)
+
+    ! Convert remaining directory separators to underscores
+    i = index(object_file,filesep)
+    do while(i > 0)
+        object_file(i:i) = '_'
+        i = index(object_file,filesep)
+    end do
+
+    ! Construct full target path
+    object_file = join_path(model%output_directory, model%package_name, &
+                        basename(object_file,suffix=.false.)//'.o')
```

Comment:
An additional remark regarding the naming of the object files.
The standard make suffix mapping will fail in case we have files with the same base name but different extensions.
```
dir/file.f90 -> dir_file.o
dir/file.f -> dir_file.o
dir/file.c -> dir_file.o
```
One could argue that having a fixed format and free format file with the same base name is bad practise and should be discouraged, but having a c-source which is wrapped in a free format file is quite intuitive, in my opinion.

I would suggest to follow the strategy of CMake and meson to just appending the object suffix:
```suggestion
    ! Construct full target path
    object_file = join_path(model%output_directory, model%package_name, &
                        object_file//'.o')
```

Reply:
Good point, I hadn't thought about that! Cheers :+1: